File: SIGAI/others/calculate.swift
//
//  calculate.swift
//  SIGAI
//
//  Created by Ainal syazwan Itamta on 22/04/2025.
//

import SwiftUI

extension ContentView {// Function to count intersections based on color
    func countIntersections() -> Int {
        var total = 0
        let pathCount = paths.count
        
        for i in 0..<pathCount {
            for j in (i + 1)..<pathCount {
                if pathsIntersect(paths[i], paths[j]) {
                    // Determine color-based intersection value
                    if (colors[i] == .red && colors[j] == .green) || (colors[i] == .green && colors[j] == .red) {
                        total += 100 // Red × Green = 100
                    } else if (colors[i] == .red && colors[j] == .blue) || (colors[i] == .blue && colors[j] == .red) {
                        total += 1000 // Red × Blue = 1000
                    } else if colors[i] == .green && colors[j] == .green {
                        total += 1 // Green × Green = 1
                    } else if colors[i] == .blue && colors[j] == .blue {
                        total += 100 // Blue × Blue = 100
                    } else if colors[i] == .red && colors[j] == .red {
                        total += 10000 // Red × Red = 10000
                    } else if colors[i] == .blue && colors[j] == .green || colors[i] == .green && colors[j] == .blue {
                        total += 10 // Blue × Green = 10 (Fix applied)
                    }
                }
            }
        }
        
        return total
    }
    
    // Function to count divisions based on color
    func countDivisions() -> Double {
        let pathCount = paths.count
        
        let totalVerticalLines: Int
        if let lockedValue = lockedVerticalLines {
            totalVerticalLines = lockedValue // Use locked value if set
        } else {
            totalVerticalLines = paths.filter { path in
                let points = extractPoints(from: path)
                guard let first = points.first, let last = points.last else { return false }
                return abs(first.x - last.x) < 30.0 // Loosen threshold for vertical detection
            }.count
            
            DispatchQueue.main.async {
                self.detectedVerticalLines = totalVerticalLines
            }
        }
        
        var divisionSum: Double = 0.0
        for i in 0..<totalVerticalLines {
            for j in 0..<pathCount {
                if i < colors.count, j < colors.count, pathsIntersect(paths[i], paths[j]) {
                    let factor = getColorMultiplier(colors[i], colors[j])
                    divisionSum += factor / Double(totalVerticalLines)
                }
            }
        }
        
        return divisionSum
    }
    
    func getColorMultiplier(_ color1: Color, _ color2: Color) -> Double {
        let red = Color.red
        let blue = Color.blue
        let green = Color.green
        let yellow = Color.yellow
        let purple = Color.purple
        
        let colorMultipliers: [Color: [Color: Double]] = [
            red:    [red: 1.0, blue: 0.1, green: 0.01, yellow: 0.001, purple: 0.0001],
            blue:   [red: 10, blue: 1.0, green: 0.1, yellow: 0.01, purple: 0.001],
            green:  [red: 100, blue: 10, green: 1.0, yellow: 0.1, purple: 0.01],
            yellow: [red: 1000, blue: 100, green: 10, yellow: 1.0, purple: 0.1],
            purple: [red: 10000, blue: 1000, green: 100, yellow: 10, purple: 1.0]
        ]
        
        return colorMultipliers[color1]?[color2] ?? 1.0
    }
    
    // Improved function to check if two paths intersect
    func pathsIntersect(_ path1: Path, _ path2: Path) -> Bool {
        let points1 = extractPoints(from: path1)
        let points2 = extractPoints(from: path2)
        
        guard points1.count > 1, points2.count > 1 else { return false }
        
        for i in 0..<(points1.count - 1) {
            for j in 0..<(points2.count - 1) {
                if checkLineIntersection(points1[i], points1[i+1], points2[j], points2[j+1]) {
                    return true
                }
            }
        }
        return false
    }
    
    // Extract points from a path
    func extractPoints(from path: Path) -> [CGPoint] {
        var points: [CGPoint] = []
        path.forEach { element in
            switch element {
            case .move(to: let point), .line(to: let point):
                points.append(point)
            default:
                break
            }
        }
        return points
    }
    
    func isVerticalLine(_ path: Path, newPoint: CGPoint?) -> Bool {
        var points = extractPoints(from: path)
        if let newPoint = newPoint {
            points.append(newPoint)
        }
        guard let first = points.first, let last = points.last else { return false }
        return abs(first.x - last.x) < 5.0 // Consider it vertical if X coordinates are close
    }
    
    // Check if two line segments intersect
    func checkLineIntersection(_ p1: CGPoint, _ p2: CGPoint, _ q1: CGPoint, _ q2: CGPoint) -> Bool {
        func cross(_ v1: CGPoint, _ v2: CGPoint) -> CGFloat {
            return v1.x * v2.y - v1.y * v2.x
        }
        
        let v1 = CGPoint(x: p2.x - p1.x, y: p2.y - p1.y)
        let v2 = CGPoint(x: q2.x - q1.x, y: q2.y - q1.y)
        let v3 = CGPoint(x: q1.x - p1.x, y: q1.y - p1.y)
        let v4 = CGPoint(x: q2.x - p1.x, y: q2.y - p1.y)
        
        let cross1 = cross(v1, v3)
        let cross2 = cross(v1, v4)
        let cross3 = cross(v2, CGPoint(x: p1.x - q1.x, y: p1.y - q1.y))
        let cross4 = cross(v2, CGPoint(x: p2.x - q1.x, y: p2.y - q1.y))
        
        return (cross1 * cross2 < 0) && (cross3 * cross4 < 0)
    }
}

File: ANDROID/sigai_flutter/lib/mainfunc/calculate.dart
import 'package:flutter/material.dart';

int countIntersections(List<List<Offset?>> paths, List<Color> colors) {
  int total = 0;
  final int pathCount = paths.length;

  for (int i = 0; i < pathCount; i++) {
    for (int j = i + 1; j < pathCount; j++) {
      if (pathsIntersect(paths[i], paths[j])) {
        final Color colorI = colors[i];
        final Color colorJ = colors[j];

        if (colorI == Colors.grey || colorJ == Colors.grey) continue;

        total += getColorMultiplier(colorI, colorJ);
      }
    }
  }

  return total;
}

double countDivisions(List<List<Offset?>> paths, List<Color> colors, {int? lockedVerticalLines}) {
  final int pathCount = paths.length;

  final verticalPathIndices = List.generate(pathCount, (i) => i)
      .where((i) {
        final path = paths[i];
        final first = path.firstWhere((p) => p != null, orElse: () => null);
        final last = path.lastWhere((p) => p != null, orElse: () => null);
        return first != null &&
            last != null &&
            (first.dx - last.dx).abs() < 30.0 &&
            colors[i] != Colors.grey;
      })
      .toList();

  final int totalVerticalLines = lockedVerticalLines ?? verticalPathIndices.length;
  if (totalVerticalLines == 0) return 0;

  double divisionSum = 0.0;

  for (final vi in verticalPathIndices) {
    for (int j = 0; j < pathCount; j++) {
      if (j < colors.length &&
        vi < colors.length &&
        vi != j &&
        pathsIntersect(paths[vi], paths[j])) {
        final cVert = colors[vi];
        final cHorz = colors[j];
        if (cVert == Colors.grey || cHorz == Colors.grey) continue;

        final vertFactor = getColorMultiplierBahagi(cVert, cVert);
        final horzFactor = getColorMultiplierBahagi(cHorz, cHorz);
        
        if (vertFactor == 0 || horzFactor == 0) continue;
        divisionSum += (horzFactor / vertFactor) / totalVerticalLines;
      }
    }
  }

  return divisionSum;
}


double getColorMultiplierBahagi(Color color1, Color color2) {
  final colorValues = {
    Colors.red: 100,
    Colors.blue: 10,
    Colors.green: 1,
    Colors.yellow: 0.1,
    Colors.purple: 0.01,
  };

  return colorValues[color1]?.toDouble() ?? 1.0;
}

int getColorMultiplier(Color color1, Color color2) {
  final red = Colors.red;
  final blue = Colors.blue;
  final green = Colors.green;

  final colorMultipliers = {
    green:  {red: 100, blue: 10, green: 1.0},
    blue: {red: 1000, blue: 100, green: 10},
    red: {red: 10000, blue: 1000, green: 100},
  };

  return colorMultipliers[color1]?[color2]?.round() ?? 1;
}

bool pathsIntersect(List<Offset?> pathA, List<Offset?> pathB) {
  final aPoints = pathA.whereType<Offset>().toList();
  final bPoints = pathB.whereType<Offset>().toList();

  if (aPoints.length < 2 || bPoints.length < 2) return false;

  for (int i = 0; i < aPoints.length - 1; i++) {
    for (int j = 0; j < bPoints.length - 1; j++) {
      if (checkLineIntersection(aPoints[i], aPoints[i + 1], bPoints[j], bPoints[j + 1])) {
        return true;
      }
    }
  }
  return false;
}

bool checkLineIntersection(Offset p1, Offset p2, Offset q1, Offset q2) {
  Offset v1 = p2 - p1;
  Offset v2 = q2 - q1;
  Offset v3 = q1 - p1;
  Offset v4 = q2 - p1;

  double cross(Offset v, Offset w) => v.dx * w.dy - v.dy * w.dx;

  final cross1 = cross(v1, v3);
  final cross2 = cross(v1, v4);
  final cross3 = cross(v2, p1 - q1);
  final cross4 = cross(v2, p2 - q1);

  bool isZero(double x) => x.abs() < 1e-5;
  return ((cross1 * cross2 <= 0) || isZero(cross1) || isZero(cross2)) &&
       ((cross3 * cross4 <= 0) || isZero(cross3) || isZero(cross4));
}
